libraries/chain/block_log.cpp

```diff
@@ -267,48 +287,73 @@ namespace eosio { namespace chain {
+      my->flush();
    }
 
-   void block_log::reset( const genesis_state& gs, const signed_block_ptr& first_block, uint32_t first_block_num ) {
-      my->close();
+   void detail::block_log_impl::flush() {
+      block_file.flush();
+      index_file.flush();
+   }
 
-      fc::remove_all( my->block_file.get_file_path() );
-      fc::remove_all( my->index_file.get_file_path() );
+   template<typename T>
+   void detail::block_log_impl::reset( const T& t, const signed_block_ptr& first_block, uint32_t first_bnum ) {
+      close();
 
-      my->reopen();
+      fc::remove_all( block_file.get_file_path() );
+      fc::remove_all( index_file.get_file_path() );
 
-      auto data = fc::raw::pack(gs);
-      my->version = 0; // version of 0 is invalid; it indicates that the genesis was not properly written to the block log
-      my->first_block_num = first_block_num;
-      my->block_file.seek_end(0);
-      my->block_file.write((char*)&my->version, sizeof(my->version));
-      my->block_file.write((char*)&my->first_block_num, sizeof(my->first_block_num));
-      my->block_file.write(data.data(), data.size());
-      my->genesis_written_to_block_log = true;
+      reopen();
+
+      version = 0; // version of 0 is invalid; it indicates that subsequent data was not properly written to the block log
+      first_block_num = first_bnum;
+
+      block_file.seek_end(0);
+      block_file.write((char*)&version, sizeof(version));
+      block_file.write((char*)&first_block_num, sizeof(first_block_num));
+
+      write(t);
+      genesis_written_to_block_log = true;
 
       // append a totem to indicate the division between blocks and header
-      auto totem = npos;
-      my->block_file.write((char*)&totem, sizeof(totem));
+      auto totem = block_log::npos;
+      block_file.write((char*)&totem, sizeof(totem));
 
       if (first_block) {
          append(first_block);
       } else {
-         my->head.reset();
-         my->head_id = {};
+         head.reset();
+         head_id = {};
       }
 
-      auto pos = my->block_file.tellp();
+      auto pos = block_file.tellp();
 
       static_assert( block_log::max_supported_version > 0, "a version number of zero is not supported" );
-      my->version = block_log::max_supported_version;
-      my->block_file.seek( 0 );
-      my->block_file.write( (char*)&my->version, sizeof(my->version) );
-      my->block_file.seek( pos );
+
+      // going back to write correct version to indicate that all block log header data writes completed successfully
+      version = block_log::max_supported_version;
+      block_file.seek( 0 );
+      block_file.write( (char*)&version, sizeof(version) );
+      block_file.seek( pos );
       flush();
    }
 
+   void block_log::reset( const genesis_state& gs, const signed_block_ptr& first_block ) {
+      my->reset(gs, first_block, 1);
+   }
+
+   void block_log::reset( const chain_id_type& chain_id, const signed_block_ptr& first_block, uint32_t first_block_num ) {
```

Comment:
Considering how `block_log::reset` was intended to be used, I think the signature of this overload should be changed to `void reset( const chain_id_type& chain_id, uint32_t first_block_num );` and then the implementation can be:
```cpp
my->reset(chain_id, signed_block_ptr(), first_block_num);
```

Of course the two calls of this function in `controller_impl::startup` will need to be updated as well.